### ISSUE TITLE
Improve accessibility of the autocomplete plugin

### ISSUE THREAD

**Issue by mlewand** [ckeditor/ckeditor4#2199]
## Type of report

Bug

## Provide detailed reproduction steps (if any)

Autocomplete accessibility markup should be reviewed. Items should gain proper roles etc.

**Comment by sculpt0r** [comment]
Some reference sources: https://w3c.github.io/aria/#combobox

**Comment by sculpt0r** [comment]
I've made some sample to easier manipulate this `combobox` - `iframe` - `listbox` relationship and test it against screen readers. Based on A11y tree: `aria-controls` works only for `id`'s in the same document. I haven't found any tip to work around it in the proper 'standard' way. I assume, we need to create our own solution for this, which leads me to some reworks on `autocomplete` plugin.

**Comment by f1ames** [comment]
> I've made some sample to easier manipulate this `combobox` - `iframe` - `listbox` relationship and test it against screen readers.

Could you share/link to those samples here?

> Based on A11y tree: `aria-controls` works only for `id`'s in the same document. I haven't found any tip to work around it in the proper 'standard' way. I assume, we need to create our own solution for this, which leads me to some reworks on `autocomplete` plugin.

:scream: 

> With some aria attributes, it works well on `divarea` plugin.  
> We can't place autocomplete list inside `wysiwygarea` plugin iframe, because it's actually our content.

Hmmm, doesn't sound like a good news for us TBH. Still :+1:  for analyzing this issue from different angles.

Maybe we can have another look on this issue too. Accessibility issue? Who Ya Gonna Call? @Comandeer!

**Comment by sculpt0r** [comment]
:see\_no\_evil: I forgot to paste a link...: [https://codepen.io/sculpt0r/pen/GRNPMpj](https://codepen.io/sculpt0r/pen/GRNPMpj) It's quite simple to change the position of `ul` list depends on what situation you want to test.

**Comment by f1ames** [comment]
This issue is partially blocked by #4052 and partially by native behavior which makes it really complex. It will require further analysis.

**Comment by f1ames** [comment]
We have revisited and analyzed this issue and all the limitation encountered during development of #4535 PR once again.

The biggest limitation here is how browsers (and AOM) and screen readers handle `iframe` elements and relation between elements in different documents (so different `iframe`s). Here, the issue is that during content edition focus is inside CKEditor 4  `iframe` while autocomplete element is in the main window. The fact that autocomplete is in the main window is the standard approach for all such elements in CKEditor 4 (like dialog, dropdowns, balloon panels, etc) which makes it usable, easier to style and possible to position over editor boundaries. The limitation here is that it's not doable to tell screen reader (which has focus inside editor `iframe`) that new element appeared in the main window. And **this is a browser / screen reader limitation which is an upstream issue we don't have control over**.

Also it turned out that #4052 could be a nice addition here but it's not a blocker for this issue (since lack of native support for such case is the blocker here).

Depends on the use case there might be possible workarounds for this issue:

* Use `divarea` editor type instead (see #).
* Try other way to insert data - autocomplete is the only element which suffers from lack of accessibility in CKEditor 4 so any other way (like using custom dialog with a richcombo/drodpown) should meet accessibility requirements.

**Comment by f1ames** [comment]
As mentioned in the [https://github.com/ckeditor/ckeditor4/issues/2199#issuecomment-818711934](https://github.com/ckeditor/ckeditor4/issues/2199#issuecomment-818711934) above there are some limitations related to iframe handling in browsers, AOM and screen readers.

The lack of accessibility in autocomplete plugin had been fixed for non-iframe based editor types in #4617 thus I'm closing this issue.
